File: SundeeFundee/Sources/SundeeFundeeKit/DomainLayer/Cycle/CycleAdaptationPolicy.swift
import Foundation

// MARK: - Types

/// Adaptation readiness tier
public enum AdaptationReadinessTier: String, Sendable {
    case low
    case neutral
    case high
}

/// Adaptation confidence level
public enum AdaptationConfidence: String, Sendable {
    case low
    case medium
    case high
}

/// Body region classification for region-specific cycle multipliers.
/// Upper and lower body respond differently to hormonal shifts.
public enum ExerciseRegion: String, Sendable, CaseIterable {
    case upper
    case lower
    case fullBody
    case core
    case conditioning
}

/// Phase-specific multipliers for cycle adaptation
public struct PhaseMultipliers: Sendable {
    public let load: Double
    public let sets: Double
    public let reps: Double

    public init(load: Double, sets: Double, reps: Double) {
        self.load = load
        self.sets = sets
        self.reps = reps
    }

    /// Linear interpolation between two multiplier sets.
    public func lerp(to other: PhaseMultipliers, t: Double) -> PhaseMultipliers {
        PhaseMultipliers(
            load: load + (other.load - load) * t,
            sets: sets + (other.sets - sets) * t,
            reps: reps + (other.reps - reps) * t
        )
    }
}

// MARK: - Constants

/// Default phase multipliers per cycle phase (used as fallback)
public let phaseMultipliers: [CyclePhase: PhaseMultipliers] = [
    .menstrual:  PhaseMultipliers(load: 0.90, sets: 0.90, reps: 0.90),
    .follicular: PhaseMultipliers(load: 1.00, sets: 1.00, reps: 1.00),
    .ovulation:  PhaseMultipliers(load: 1.12, sets: 1.05, reps: 0.95),
    .luteal:     PhaseMultipliers(load: 0.97, sets: 0.95, reps: 0.92),
]

/// Region-specific overrides — upper and lower body respond differently to hormonal shifts.
/// Lower body gets bigger menstrual reduction and bigger ovulation boost.
public let regionPhaseMultipliers: [ExerciseRegion: [CyclePhase: PhaseMultipliers]] = [
    .lower: [
        .menstrual:  PhaseMultipliers(load: 0.85, sets: 0.85, reps: 0.88),
        .follicular: PhaseMultipliers(load: 1.00, sets: 1.00, reps: 1.00),
        .ovulation:  PhaseMultipliers(load: 1.15, sets: 1.08, reps: 0.93),
        .luteal:     PhaseMultipliers(load: 0.95, sets: 0.93, reps: 0.90),
    ],
    .upper: [
        .menstrual:  PhaseMultipliers(load: 0.92, sets: 0.92, reps: 0.92),
        .follicular: PhaseMultipliers(load: 1.00, sets: 1.00, reps: 1.00),
        .ovulation:  PhaseMultipliers(load: 1.10, sets: 1.03, reps: 0.96),
        .luteal:     PhaseMultipliers(load: 0.98, sets: 0.96, reps: 0.93),
    ],
]

private let readinessScales: [AdaptationReadinessTier: Double] = [
    .low:     0.6,
    .neutral: 1.0,
    .high:    1.2,
]

private let confidenceScales: [AdaptationConfidence: Double] = [
    .low:    0.55,
    .medium: 0.8,
    .high:   1.0,
]

// MARK: - Helpers

private func clamp(_ value: Double, min: Double, max: Double) -> Double {
    Swift.min(Swift.max(value, min), max)
}

// MARK: - Exercise Region Classification

/// Classify an exercise name into a body region for region-specific multipliers.
public func classifyExerciseRegion(_ exerciseName: String) -> ExerciseRegion {
    let lower = exerciseName.lowercased()

    let lowerKeywords = ["squat", "deadlift", "leg", "hip thrust", "lunge",
                         "calf", "glute", "hamstring", "good morning"]
    if lowerKeywords.contains(where: { lower.contains($0) }) { return .lower }

    let upperKeywords = ["bench", "press", "row", "pull-up", "chin-up", "curl",
                         "dip", "lat pull", "face pull", "fly", "raise",
                         "overhead", "shoulder", "tricep"]
    if upperKeywords.contains(where: { lower.contains($0) }) { return .upper }

    let coreKeywords = ["plank", "sit-up", "crunch", "ab ", "core", "hollow",
                        "l-sit", "v-up", "toes-to-bar"]
    if coreKeywords.contains(where: { lower.contains($0) }) { return .core }

    let condKeywords = ["run", "row ", "bike", "swim", "burpee", "wall ball",
                        "box jump", "double under", "skierg"]
    if condKeywords.contains(where: { lower.contains($0) }) { return .conditioning }

    // Clean, snatch, jerk, carry, farmer's walk → full body
    return .fullBody
}

/// Look up the phase multipliers for a given phase and optional exercise region.
/// Falls back to the default (non-region) multipliers when no region override exists.
public func resolvePhaseMultipliers(
    phase: CyclePhase,
    region: ExerciseRegion? = nil
) -> PhaseMultipliers {
    if let region, let regionMap = regionPhaseMultipliers[region],
       let multipliers = regionMap[phase] {
        return multipliers
    }
    return phaseMultipliers[phase] ?? PhaseMultipliers(load: 1.0, sets: 1.0, reps: 1.0)
}

// MARK: - Phase Transition Blending

/// Returns blended multipliers when near a phase boundary (within 1 day).
/// Prevents jarring jumps between phases — e.g., follicular→ovulation blends
/// at the transition day rather than switching abruptly.
public func blendedMultiplier(
    phase: CyclePhase,
    cycleDay: Int,
    settings: CycleSettings,
    region: ExerciseRegion? = nil
) -> PhaseMultipliers {
    let boundaries = getPhaseBoundaries(settings: settings)
    let currentMultipliers = resolvePhaseMultipliers(phase: phase, region: region)

    guard let boundary = boundaries[phase] else { return currentMultipliers }

    // Check if we're within 1 day of the END of current phase (transition to next)
    if cycleDay >= boundary.end {
        let nextPhase = nextCyclePhase(after: phase)
        let nextMultipliers = resolvePhaseMultipliers(phase: nextPhase, region: region)
        // Blend: 50% current + 50% next at the exact boundary day
        return currentMultipliers.lerp(to: nextMultipliers, t: 0.5)
    }

    // Check if we're at the START of current phase (transition from previous)
    if cycleDay <= boundary.start {
        let prevPhase = previousCyclePhase(before: phase)
        let prevMultipliers = resolvePhaseMultipliers(phase: prevPhase, region: region)
        return prevMultipliers.lerp(to: currentMultipliers, t: 0.5)
    }

    return currentMultipliers
}

/// Next phase in the cycle sequence
private func nextCyclePhase(after phase: CyclePhase) -> CyclePhase {
    switch phase {
    case .menstrual:  return .follicular
    case .follicular: return .ovulation
    case .ovulation:  return .luteal
    case .luteal:     return .menstrual
    }
}

/// Previous phase in the cycle sequence
private func previousCyclePhase(before phase: CyclePhase) -> CyclePhase {
    switch phase {
    case .menstrual:  return .luteal
    case .follicular: return .menstrual
    case .ovulation:  return .follicular
    case .luteal:     return .ovulation
    }
}

// MARK: - Public API

/// Resolve readiness tier from a 0-10 score
public func resolveReadinessTier(score: Double?) -> AdaptationReadinessTier {
    guard let score else { return .neutral }
    if score <= 3 { return .low }
    if score >= 8 { return .high }
    return .neutral
}

/// Resolve confidence level from cycle data availability
public func resolveConfidence(
    currentPhase: CyclePhase?,
    lastKnownPhase: CyclePhase?,
    periodLogCount: Int,
    lastPeriodStart: Date?,
    referenceDate: Date = Date()
) -> AdaptationConfidence {
    if periodLogCount == 0 { return .low }

    if let lastPeriodStart {
        let daysSince = referenceDate.timeIntervalSince(lastPeriodStart) / (60 * 60 * 24)
        if daysSince > 60 { return .medium }
    }

    if currentPhase != nil && periodLogCount >= 3 { return .high }
    if currentPhase != nil { return .medium }
    return .low
}

/// Apply cycle phase adjustment to an exercise's sets, reps, and load.
/// Optionally accepts an exercise region for region-specific multipliers.
public func applyPhaseAdjustment(
    sets: ExerciseValue,
    reps: ExerciseValue,
    percent1RM: Double?,
    phase: CyclePhase,
    readinessTier: AdaptationReadinessTier,
    confidence: AdaptationConfidence,
    exerciseRegion: ExerciseRegion? = nil
) -> (sets: ExerciseValue, reps: ExerciseValue, percent1RM: Double?) {
    let settings = resolvePhaseMultipliers(phase: phase, region: exerciseRegion)
    let rs = readinessScales[readinessTier] ?? 1.0
    let cs = confidenceScales[confidence] ?? 1.0

    func blendMult(_ target: Double) -> Double {
        clamp(1.0 + (target - 1.0) * rs * cs, min: 0.75, max: 1.25)
    }

    let adjustedSets = adjustExerciseValueByMultiplier(sets, multiplier: blendMult(settings.sets))
    let adjustedReps = adjustExerciseValueByMultiplier(reps, multiplier: blendMult(settings.reps))
    let adjustedLoad = percent1RM.map { clamp($0 * blendMult(settings.load), min: 0.4, max: 1.1) }

    return (sets: adjustedSets, reps: adjustedReps, percent1RM: adjustedLoad)
}

File: android/core/src/main/java/com/sundeefundee/core/domain/cycle/CycleAdaptationPolicy.kt
package com.sundeefundee.core.domain.cycle

import com.sundeefundee.core.domain.exercise.ExerciseValue
import com.sundeefundee.core.model.CyclePhase
import com.sundeefundee.core.model.CycleSettings
import kotlinx.datetime.LocalDate

// Adaptation readiness tier
enum class AdaptationReadinessTier { LOW, NEUTRAL, HIGH }

// Adaptation confidence level
enum class AdaptationConfidence { LOW, MEDIUM, HIGH }

// Body region classification for region-specific multipliers
enum class ExerciseRegion {
    UPPER, LOWER, FULL_BODY, CORE, CONDITIONING
}

// Phase-specific multipliers for cycle adaptation
data class PhaseMultipliers(
    val load: Double,
    val sets: Double,
    val reps: Double
) {
    fun lerp(other: PhaseMultipliers, t: Double): PhaseMultipliers = PhaseMultipliers(
        load = load + (other.load - load) * t,
        sets = sets + (other.sets - sets) * t,
        reps = reps + (other.reps - reps) * t
    )
}

// Default phase multipliers
val PHASE_MULTIPLIERS = mapOf(
    CyclePhase.MENSTRUAL to PhaseMultipliers(load = 0.90, sets = 0.90, reps = 0.90),
    CyclePhase.FOLLICULAR to PhaseMultipliers(load = 1.00, sets = 1.00, reps = 1.00),
    CyclePhase.OVULATION to PhaseMultipliers(load = 1.12, sets = 1.05, reps = 0.95),
    CyclePhase.LUTEAL to PhaseMultipliers(load = 0.97, sets = 0.95, reps = 0.92)
)

// Region-specific overrides
val REGION_PHASE_MULTIPLIERS = mapOf(
    ExerciseRegion.LOWER to mapOf(
        CyclePhase.MENSTRUAL to PhaseMultipliers(load = 0.85, sets = 0.85, reps = 0.88),
        CyclePhase.FOLLICULAR to PhaseMultipliers(load = 1.00, sets = 1.00, reps = 1.00),
        CyclePhase.OVULATION to PhaseMultipliers(load = 1.15, sets = 1.08, reps = 0.93),
        CyclePhase.LUTEAL to PhaseMultipliers(load = 0.95, sets = 0.93, reps = 0.90)
    ),
    ExerciseRegion.UPPER to mapOf(
        CyclePhase.MENSTRUAL to PhaseMultipliers(load = 0.92, sets = 0.92, reps = 0.92),
        CyclePhase.FOLLICULAR to PhaseMultipliers(load = 1.00, sets = 1.00, reps = 1.00),
        CyclePhase.OVULATION to PhaseMultipliers(load = 1.10, sets = 1.03, reps = 0.96),
        CyclePhase.LUTEAL to PhaseMultipliers(load = 0.98, sets = 0.96, reps = 0.93)
    )
)

private val READINESS_SCALES = mapOf(
    AdaptationReadinessTier.LOW to 0.6,
    AdaptationReadinessTier.NEUTRAL to 1.0,
    AdaptationReadinessTier.HIGH to 1.2
)

private val CONFIDENCE_SCALES = mapOf(
    AdaptationConfidence.LOW to 0.55,
    AdaptationConfidence.MEDIUM to 0.8,
    AdaptationConfidence.HIGH to 1.0
)

private fun clamp(value: Double, min: Double, max: Double): Double =
    value.coerceIn(min, max)

// Classify an exercise name into a body region
fun classifyExerciseRegion(exerciseName: String): ExerciseRegion {
    val lower = exerciseName.lowercase()

    val lowerKeywords = listOf("squat", "deadlift", "leg", "hip thrust", "lunge", "calf", "glute", "hamstring", "good morning")
    if (lowerKeywords.any { lower.contains(it) }) return ExerciseRegion.LOWER

    val upperKeywords = listOf("bench", "press", "row", "pull-up", "chin-up", "curl", "dip", "lat pull", "face pull", "fly", "raise", "overhead", "shoulder", "tricep")
    if (upperKeywords.any { lower.contains(it) }) return ExerciseRegion.UPPER

    val coreKeywords = listOf("plank", "sit-up", "crunch", "ab ", "core", "hollow", "l-sit", "v-up", "toes-to-bar")
    if (coreKeywords.any { lower.contains(it) }) return ExerciseRegion.CORE

    val condKeywords = listOf("run", "row ", "bike", "swim", "burpee", "wall ball", "box jump", "double under", "skierg")
    if (condKeywords.any { lower.contains(it) }) return ExerciseRegion.CONDITIONING

    return ExerciseRegion.FULL_BODY
}

// Look up phase multipliers for a given phase and optional region
fun resolvePhaseMultipliers(phase: CyclePhase, region: ExerciseRegion? = null): PhaseMultipliers {
    if (region != null) {
        val regionMap = REGION_PHASE_MULTIPLIERS[region]
        if (regionMap != null) {
            return regionMap[phase] ?: PHASE_MULTIPLIERS[phase] ?: PhaseMultipliers(1.0, 1.0, 1.0)
        }
    }
    return PHASE_MULTIPLIERS[phase] ?: PhaseMultipliers(1.0, 1.0, 1.0)
}

// Next phase in cycle sequence
private fun nextCyclePhase(phase: CyclePhase): CyclePhase = when (phase) {
    CyclePhase.MENSTRUAL -> CyclePhase.FOLLICULAR
    CyclePhase.FOLLICULAR -> CyclePhase.OVULATION
    CyclePhase.OVULATION -> CyclePhase.LUTEAL
    CyclePhase.LUTEAL -> CyclePhase.MENSTRUAL
}

// Previous phase in cycle sequence
private fun previousCyclePhase(phase: CyclePhase): CyclePhase = when (phase) {
    CyclePhase.MENSTRUAL -> CyclePhase.LUTEAL
    CyclePhase.FOLLICULAR -> CyclePhase.MENSTRUAL
    CyclePhase.OVULATION -> CyclePhase.FOLLICULAR
    CyclePhase.LUTEAL -> CyclePhase.OVULATION
}

// Blended multipliers at phase boundaries (within 1 day)
fun blendedMultiplier(
    phase: CyclePhase,
    cycleDay: Int,
    settings: CycleSettings,
    region: ExerciseRegion? = null
): PhaseMultipliers {
    val boundaries = getPhaseBoundaries(settings)
    val current = resolvePhaseMultipliers(phase, region)
    val boundary = boundaries[phase] ?: return current

    if (cycleDay >= boundary.end) {
        val next = resolvePhaseMultipliers(nextCyclePhase(phase), region)
        return current.lerp(next, 0.5)
    }

    if (cycleDay <= boundary.start) {
        val prev = resolvePhaseMultipliers(previousCyclePhase(phase), region)
        return prev.lerp(current, 0.5)
    }

    return current
}

// Resolve readiness tier from 0-10 score
fun resolveReadinessTier(score: Double?): AdaptationReadinessTier {
    score ?: return AdaptationReadinessTier.NEUTRAL
    return when {
        score <= 3 -> AdaptationReadinessTier.LOW
        score >= 8 -> AdaptationReadinessTier.HIGH
        else -> AdaptationReadinessTier.NEUTRAL
    }
}

// Resolve confidence from cycle data availability
fun resolveConfidence(
    currentPhase: CyclePhase?,
    periodLogCount: Int,
    lastPeriodStartDate: LocalDate?,
    referenceDate: LocalDate
): AdaptationConfidence {
    if (periodLogCount == 0) return AdaptationConfidence.LOW

    if (lastPeriodStartDate != null) {
        val daysSince = referenceDate.toEpochDays() - lastPeriodStartDate.toEpochDays()
        if (daysSince > 60) return AdaptationConfidence.MEDIUM
    }

    if (currentPhase != null && periodLogCount >= 3) return AdaptationConfidence.HIGH
    if (currentPhase != null) return AdaptationConfidence.MEDIUM
    return AdaptationConfidence.LOW
}

// Adjust an ExerciseValue by a multiplier
fun adjustExerciseValueByMultiplier(value: ExerciseValue, multiplier: Double): ExerciseValue = when (value) {
    is ExerciseValue.Fixed -> ExerciseValue.Fixed(maxOf(1, (value.value * multiplier).toInt()))
    is ExerciseValue.Amrap -> value
    is ExerciseValue.Range -> ExerciseValue.Range(
        min = maxOf(1, (value.min * multiplier).toInt()),
        max = maxOf(1, (value.max * multiplier).toInt()).also { max ->
            if (max < maxOf(1, (value.min * multiplier).toInt())) {
                return ExerciseValue.Range(min = max, max = maxOf(1, (value.min * multiplier).toInt()))
            }
        }
    )
    is ExerciseValue.Text -> value
}

// Convenience: adjust weight by phase (returns adjusted weight)
fun adjustWeightByPhase(weight: Double, phase: CyclePhase): Double {
    val multiplier = PHASE_MULTIPLIERS[phase]?.load ?: 1.0
    return weight * multiplier
}

// Apply cycle phase adjustment to sets, reps, and load
fun applyPhaseAdjustment(
    sets: ExerciseValue,
    reps: ExerciseValue,
    percent1RM: Double?,
    phase: CyclePhase,
    readinessTier: AdaptationReadinessTier,
    confidence: AdaptationConfidence,
    exerciseRegion: ExerciseRegion? = null
): Triple<ExerciseValue, ExerciseValue, Double?> {
    val settings = resolvePhaseMultipliers(phase, exerciseRegion)
    val rs = READINESS_SCALES[readinessTier] ?: 1.0
    val cs = CONFIDENCE_SCALES[confidence] ?: 1.0

    fun blendMult(target: Double): Double = clamp(1.0 + (target - 1.0) * rs * cs, 0.75, 1.25)

    val adjustedSets = adjustExerciseValueByMultiplier(sets, blendMult(settings.sets))
    val adjustedReps = adjustExerciseValueByMultiplier(reps, blendMult(settings.reps))
    val adjustedLoad = percent1RM?.let { clamp(it * blendMult(settings.load), 0.4, 1.1) }

    return Triple(adjustedSets, adjustedReps, adjustedLoad)
}
